Add Baltop API

JRoy opened the pull request:
This moves storage of balances from the baltop command into the UserMap. This was needed by Glare to able to get a hold of all users balances without causing jvm hell on the usermap.

To access this API as an end user;
```java
import net.essentialsx.api.v2.services.BalanceTop;
//...
BalanceTop api = Bukkit.getServer().getServicesManager().load(BalanceTop.class);
```

Closes #3100
Closes #3540

JRoy committed:
Add balancetop api

JRoy committed:
Move API to its own class

JRoy committed:
Reduce diff in UserMap

JRoy committed:
Improve API + Fix startup bug

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api

# Conflicts:
#	Essentials/src/com/earth2me/essentials/commands/Commandbalancetop.java

JRoy committed:
Glare.mp4

darbyjack reviewed (dismissed):
I've reviewed this PR and deem it to be worthy of merging. Thank you for taking the time to work with me and implementing this feature. This should be a big help to anyone who uses the Essentials api.

mbax commented on Essentials/src/com/earth2me/essentials/commands/Commandbalancetop.java:
```diff
@@ -102,38 +98,17 @@ public void run() {
+                    cache.getLines().add(tl("serverTotal", NumberUtil.displayCurrency(ess.getBalanceTop().getBalanceTopTotal(), ess)));
                     int pos = 1;
-                    for (final Map.Entry<String, BigDecimal> entry : sortedEntries) {
+                    for (Map.Entry<String, BigDecimal> entry : ess.getBalanceTop().getBalanceTopCache()) {
```
Why remove `final` here?

JRoy replied:
conflict hell

mbax commented on Essentials/src/com/earth2me/essentials/BalanceTop.java:
```diff
@@ -0,0 +1,92 @@
+
+    /**
+     * Returns a future which will be completed with the latest calculated balance top map.
+     * @return A future with the latest balance top map.
```
Should there be a linebreak before the `@return` here?

JRoy committed:
mbax.ogg

JRoy commented:
![](https://github.trello.services/images/mini-trello-icon.png) [New BalanceTop API](https://trello.com/c/nVhEhepa/2-new-balancetop-api)

JRoy commented:
![](https://github.trello.services/images/mini-trello-icon.png) [New BalanceTop API](https://trello.com/c/nVhEhepa/2-new-balancetop-api)

mdcfe commented on Essentials/src/com/earth2me/essentials/BalanceTop.java:
```diff
@@ -0,0 +1,93 @@
+        calculateBalanceTopMap();
+    }
+
+    protected void calculateBalanceTopMap() {
```
IMO this should be private to `BalanceTop`, and not exposed to plugins *or* `/balancetop` - both should wait on the async method.

mdcfe commented on Essentials/src/com/earth2me/essentials/BalanceTop.java:
```diff
@@ -0,0 +1,93 @@
+     *
+     * @return A future with the latest balance top map.
+     */
+    public CompletableFuture<List<Map.Entry<String, BigDecimal>>> calculateBalanceTopMapAsync() {
```
Returning the results here makes it less clear to API callers what this does. I'd suggest that this method wraps `calculateBalanceTopMap` and returns a `CompletableFuture<Void>` that completes when done, then plugins can access the newly-updated cached baltop map from `getBalanceTopCache`.

mdcfe commented on Essentials/src/com/earth2me/essentials/BalanceTop.java:
```diff
@@ -0,0 +1,93 @@
+
+    public BigDecimal getBalanceTopTotal() {
+        return balanceTopTotal;
+    }
```
Consider adding a timestamp that stores the last update time, so that plugins and `/baltop` can decide to update the cache based on when the last update occurred?

mdcfe commented on Essentials/src/com/earth2me/essentials/CommandSource.java:
```diff
@@ -53,4 +53,11 @@ public String getSelfSelector() {
+     */
+    public void calculateBalanceTopMap(IEssentials ess) {
+        ess.getBalanceTop().calculateBalanceTopMap();
+    }
```
```suggestion
```

This is leaking way too far beyond where it should exist

mdcfe commented on Essentials/src/com/earth2me/essentials/Essentials.java:
```diff
@@ -232,6 +234,9 @@ public void onEnable() {
                 execTimer.mark("Init(Usermap)");
 
+                balanceTop = new BalanceTop(this);
+                execTimer.mark("Init(Balancetop)");
```
```suggestion
                execTimer.mark("Init(BalanceTop)");
```

pop4959 reviewed (changes requested):
In addition to MD's comments, I think we should make an effort to ensure that plugins aren't able to spam this by requesting a freshly computed baltop too often, or simultaneously.

My suggested solution would be to only allow baltop to be computed once at any given time, and if a plugin requests it again before the previous one was complete, or at the same time as another plugin, the same or an identical future will be created and completed.

This also has the added benefit of overall shorter wait times if this API is being heavily used. For example, if baltop takes 2 minutes to compute, and plugin A requests a minute before plugin B, plugin B will receive the same data in half the time of plugin A.

JRoy committed:
Review changes

Commandbalancetop now just implements the BalanceTop API
rather than being granted special permissions from it

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api

darbyjack reviewed (dismissed):
Looks good to me.

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api

JRoy committed:
Fix javadoc formatting

JRoy committed:
Refactor to new api package

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/Essentials.java:
```diff
@@ -947,6 +952,11 @@ public UserMap getUserMap() {
     }
 
+    @Override
+    public BalanceTopImpl getBalanceTop() {
```
Best to return `BalanceTop` here so people using `Essentials` will import the API class rather than impl

mdcfe commented on Essentials/src/main/java/net/essentialsx/api/v2/services/BalanceTop.java:
```diff
@@ -0,0 +1,64 @@
+
+/**
+ * A class which provides numerous methods to interact with Essentials' balance top calculations.
+ *
```
```suggestion
 * <p>
```

JRoy committed:
We can't have nice things

Proximyst commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java:
```diff
@@ -54,25 +48,17 @@ protected void run(final Server server, final CommandSource sender, final String
+            return;
+        }
+
+        //If there are less than 50 users in our usermap, there is no need to display a warning as these calculations should be done quickly
```
```suggestion
        // If there are less than 50 users in our usermap, there is no need to display a warning as these calculations should be done quickly
```

Proximyst commented on Essentials/src/main/java/com/earth2me/essentials/BalanceTopImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+public class BalanceTopImpl implements BalanceTop {
+    private final IEssentials ess;
+    private List<Map.Entry<String, BigDecimal>> topCache = new ArrayList<>();
```
Guava anyone ?

TehBrian replied:
absolutely not

JRoy replied:
don't need any of guava's cache features. the whole list is thrown out periodically anyway.

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java

Co-authored-by: Mariell <proximyst@proximyst.com>

TehBrian commented on Essentials/src/main/java/com/earth2me/essentials/Essentials.java:
```diff
@@ -125,6 +126,7 @@
     private transient PermissionsHandler permissionsHandler;
     private transient AlternativeCommandsHandler alternativeCommandsHandler;
     private transient UserMap userMap;
+    private transient BalanceTopImpl balanceTop;
```
shouldn't the type be the interface and not the impl?

JRoy replied:
`Essentials` is an impli class itself so it doesn't really matter. I'll keep it in case we need to use any impli methods in the Ess class itself.

JRoy committed:
Register with service api

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api

JRoy committed:
Add fake service manager for tests

JRoy committed:
Cache baltop exempt state in userdata

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/BalanceTopImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+public class BalanceTopImpl implements BalanceTop {
+    private final IEssentials ess;
+    private List<Map.Entry<String, BigDecimal>> topCache = new ArrayList<>();
```
`LinkedHashMap` is probably what we're after?

mdcfe commented on Essentials/src/main/java/net/essentialsx/api/v2/services/BalanceTop.java:
```diff
@@ -0,0 +1,64 @@
+     * @see BalanceTop#calculateBalanceTopMapAsync()
+     * @return The balance top cache.
+     */
+    List<Map.Entry<String, BigDecimal>> getBalanceTopCache();
```
```suggestion
    LinkedHashMap<String, BigDecimal> getBalanceTopCache();
```

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api

JRoy committed:
Switch to a less cancer storage method

Hellcat45 commented:
The Reason why my Player was not Show in Baltop. Same Username with Full UUID but was an NPC from Citizen. Hope that Info is Helpful.

npc: true
lastAccountName: jaminez
money: '1000'

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api

# Conflicts:
#	Essentials/src/main/java/net/essentialsx/api/v2/package-info.java
#	Essentials/src/main/java/net/essentialsx/api/v2/services/package-info.java

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java:
```diff
@@ -88,89 +75,42 @@ protected void run(final Server server, final CommandSource sender, final String
+
+            final boolean fresh = force || ess.getBalanceTop().isCacheLocked() || ess.getBalanceTop().getCacheAge() <= System.currentTimeMillis() - CACHETIME;
+            final CompletableFuture<Void> future = fresh ? ess.getBalanceTop().calculateBalanceTopMapAsync() : CompletableFuture.completedFuture(null);
+            future.thenAccept(unused -> {
```
```suggestion
            future.thenRun(() -> {
```

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java

Co-authored-by: MD <1917406+mdcfe@users.noreply.github.com>

JRoy committed:
Merge branch '2.x' into feature/improved-baltop-api
